QM flip-link

Hoverbear opened the pull request:
Building off the excellent work in #431 , this adds docs and testing for `flip-link` to bring it to a point we can call it Quality Managed.

We'll need to follow up with some PR establishing the term in our docs.

## Reviewing

Please ensure you see `flip-link` in the job the CI runs! You can check locally with:

```bash
BOOTSTRAP_TRACING=info ./x.py --stage 2 test $(ferrocene/ci/split-tasks.py test) -vvv --dry-run
```

Hoverbear committed:
QM flip-link

Hoverbear committed:
Add flip-link test step

jyn514 reviewed (commented):
nice work, this is a lot simpler than i thought it would be. 9/10 some notes

jyn514 commented on ferrocene/doc/user-manual/src/flip-link/usage.rst:
```diff
@@ -0,0 +1,63 @@
+.. code-block:: bash
+
+   export RUSTC_LOG=rustc_codegen_ssa::back::link=info
+   export RUST_LOG=info
```
i don't love this, opened an upstream issue https://github.com/knurling-rs/flip-link/issues/124

jyn514 reviewed (approved):
Let's Get This Merged

Hoverbear commented:
bors merge

jyn514 committed:
bootstrap: Make Rustflags public

This may or may not be something upstream is interested in merging.
We need it because we're doing "weird" things with the nested `cargo
test` invocation in flip-link.

jyn514 committed:
bootstrap: Unconditionally register a thumbv7 target

It's really quite unfortunate that we need to do this. The problem is
such:

- Bootstrap creates a list of all targets up-front in `cc_detect` from
  the `--target` CLI argument. Once created, this list is immutable.
- The flip-link tests run `cargo build --target=thumbv7`
  unconditionally. So if we test flip-link, we need thumb (otherwise
  bootstrap crashes when building libstd).
- Bootstrap has no way of determining whether a test *will* run. The
  current logic is in `StepDescription::run` and does not return data,
  it directly executes the steps. So we can't check in `cc_detect`
  whether we'll run flip-link tests.

As a hack, unconditionally require the thumbv7 target in `cc_detect`.
On my machine, this adds several new warnings when compiling:
```
cargo:warning=Compiler family detection failed due to error:
ToolNotFound: failed to find tool "arm-none-eabi-gcc": No such file or
directory (os error 2)
cargo:warning=Compiler family detection failed due to error:
ToolNotFound: failed to find tool "arm-none-eabi-g++": No such file or
directory (os error 2)
cargo:warning=Compiler family detection failed due to error:
ToolNotFound: failed to find tool "arm-none-eabi-g++": No such file or
directory (os error 2)
```
I'm not sure how to avoid this. Open to suggestions.

jyn514 committed:
bootstrap: flip-link is a target tool, not a bootstrap tool

In particular, we distribute it, which means it should be possible to
compile with stage2. This also allows us to cross-compile a flip-link
from e.g. a Darwin host to a Linux target; that's totally fine.

This commit isn't strictly necessary to run flip-link tests, it just
stood out to me while working on them.

jyn514 committed:
bootstrap: Test flip-link

- Remove incorrect `run.target`. We always test on the host, so there's
  nothing to vary here.
- Build the host target before compiling flip-link
- Build the thumbv7 standard library before running integration tests
- Set PATH to point to the newly built flip-link
- Compile the test binaries separately from running them so that they
  aren't affected by modified RUSTFLAGS
- Update bootstrap snapshot tests

pvdrz reviewed (approved):
LGTM, bors merge

tshepang commented on src/bootstrap/src/core/builder/tests.rs:
```diff
@@ -809,6 +815,7 @@ mod snapshot {
         [build] rustc 1 <host> -> LldWrapper 2 <target1>
         [build] rustc 1 <host> -> LlvmBitcodeLinker 2 <target1>
         [build] rustdoc 2 <target1>
+        [build] rustc 2 <host> -> std 2 <target1>
```
similar

pvdrz reviewed (approved):
bors merge

jyn514 committed:
fix bad rebase

Hoverbear committed:
Adjust to new upstream changes

Hoverbear committed:
Remove some unnecessary changes

Hoverbear committed:
Tidy some extraneous changes

Hoverbear committed:
It's possible to now just run cargo test for flip-link

Hoverbear committed:
Add overview page to docs

Hoverbear commented on src/bootstrap/src/core/builder/tests.rs:
```diff
@@ -563,6 +563,7 @@ mod snapshot {
         [build] rustc 0 <host> -> rustc 1 <host>
         [build] rustc 1 <host> -> std 1 <host>
         [build] rustdoc 1 <host>
+        [build] rustc 1 <host> -> std 1 <host>
```
It is indeed rather odd...

Hoverbear replied:
I spent some time looking at these and I believe they're caused by the `flip-link` tool being `is_default_step(...) { true }` I do not believe they indicate a problem.

Hoverbear committed:
Try to fix the FIXME by making a new test job

jyn514 commented on ferrocene/doc/user-manual/src/flip-link/usage.rst:
```diff
@@ -0,0 +1,63 @@
+
+.. code-block:: bash
+
+   export RUSTC_LOG=rustc_codegen_ssa::back::link=info
```
starting with https://github.com/ferrocene/ferrocene/pull/2370 these will be shown by default. I think you can remove RUSTC_LOG here.

jyn514 commented on src/bootstrap/src/ferrocene/test.rs:
```diff
@@ -203,3 +205,48 @@ impl Step for DiffUpstream {
+        builder.ensure(Std::new(compiler, host));
+        // The flip link tests require a thumbv7em-none-eabi target to exist
+        let thumb = TargetSelection::from_user("thumbv7em-none-eabi");
+        builder.ensure(Std::new(compiler, thumb));
```
I wonder if, rather than hard-coding this, we could add `builder.targets.contains(thumbv7)` as a condition to `is_default_step`? then we have *way* fewer hacks in bootstrap while still not surfacing useless build errors.

Hoverbear replied:
So after removing this, the tests started failing! 

```
$ ./x.py test flip-link --target thumbv7em-none-eabi
...
error[E0463]: can\'t find crate for `core`
  |
  = note: the `thumbv7em-none-eabi` target may not be installed
  = help: consider adding the standard library to the sysroot with `x build library --target thumbv7em-none-eabi`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

For more information about this error, try `rustc --explain E0463`.
error: could not compile `nb` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `vcell` (lib) due to 1 previous error
error: could not compile `bare-metal` (lib) due to 1 previous error
error: could not compile `bitfield` (lib) due to 1 previous error
error: could not compile `cortex-m-rt` (lib) due to 1 previous error
error: could not compile `void` (lib) due to 1 previous error


failures:
    should_link_example_firmware::case_1_normal
    should_link_example_firmware::case_2_custom_linkerscript
    should_verify_memory_layout
```

jyn514 replied:
Right, you still need to `ensure` the target, but you're getting it from `run.targets`, not hard-coding it.

Hoverbear committed:
Add release note

Hoverbear committed:
Check for thumbv7em in is_default_step

jyn514 commented on ferrocene/doc/user-manual/src/flip-link/usage.rst:
```diff
@@ -0,0 +1,62 @@
+-------
+
+If you want to see what ``flip-link`` is up to, you can set these environment variables:
+
+.. code-block:: bash
+
+   export RUST_LOG=info
+
+
+This will produce something like:
+
+.. code-block:: bash
+
+   $ cargo build
+   ...
+   INFO rustc_codegen_ssa::back::link linker stderr:
+   [INFO  flip_link] found MemoryEntry(line=3, origin=0x20000000, length=0x10000) in ./target/thumbv7em-none-eabi/debug/build/lm3s6965-3b7087c63b161e04/out/memory.x
+   [INFO  flip_link] used RAM spans: origin=0x20000000, length=12, align=4
+   [INFO  flip_link] new RAM region: ORIGIN=0x2000fff0, LENGTH=16
+   INFO rustc_codegen_ssa::back::link linker stdout:
+   INFO rustc_codegen_ssa::back::link linker stdout:
+      Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.08s
+
+You can see even more detail about how expressions are parsed using ``RUST_LOG=debug``.
```
can you regenerate this with latest nightly (of upstream), please? it looks much different now.

Hoverbear committed:
Move into own file, restore dependency on thumbv7em std

Hoverbear committed:
Update flip-link example output
